# AfterWork Data Science: Data Wrangling with Python Project

## 1. Defining the Question

### a) Specifying the Data Analysis Question

What are the recommendations that can be provided, on how to reduce the cases of breakdown of buses in the city of New York. 

> i.e. Which was the most improved country year 2017 - 2018 as per the GDP ranking?

### b) Defining the Metric for Success

To provide a solution for the problem, we first have to investigate the causes of breakdown from provided data. Hence the solution of the problem is a success when, if impleted, the data for the breakdown should be notice a change. 

### c) Understanding the context 

The dataset provided comes from the Bus Breakdown and Delay system which collects information from school bus vendors operating out in the field in real time. Bus staff that encounter delays during the route are instructed to radio the dispatcher at the bus vendor’s central office. The bus vendor staff are then instructed to log into the Bus Breakdown and Delay system to record the event and notify OPT. OPT customer service agents use this system to inform parents who call with questions regarding bus service. The Bus Breakdown and Delay system is publicly accessible and contains real time updates. All information in the system is entered by school bus vendor staff

### d) Recording the Experimental Design

**Steps taken to answer the given question.**
The obtained dataset is to be; read, explored, cleaned and later onwards performing an analysis on it.

In this case, emphasis is taken for data wrangling, where we try to answer the below questions;
1. Which bus companies that had the highest breakdowns?
2. What were the top 3 reasons for bus delays?
3. How many students were in the buses when they broke down? 
4. Which were most frequent reasons for bus breakdowns?
5. What were the most frequent reasons for the bus running late? 
6. What was the average delay time of each reason type?

### e) Data Relevance

The dataset contains details anaylsis of buses involved, containing the type of buses, there companies, number of students in each bus, reasons for delay, delay time, e.tc.
The dataset is also recorded consistently, hene providing a huge data information to carryout the analysis.

## 2. Reading the Data

In [ ]:
# Importing our libraries
# ---
#

import pandas as pd

import numpy as np

In [ ]:
# Load the data below
# --- 
# Dataset url = https://bit.ly/BusBreakdownDataset
# --- 
# 
bb = pd.read_csv("https://bit.ly/BusBreakdownDataset")

In [ ]:
# Checking the first 5 rows of data
# ---
#
bb.head()

School_Year  Busbreakdown_ID  ... Breakdown_or_Running_Late School_Age_or_PreK
0   2015-2016          1227538  ...              Running Late         School-Age
1   2015-2016          1227539  ...              Running Late         School-Age
2   2015-2016          1227540  ...              Running Late              Pre-K
3   2015-2016          1227541  ...              Running Late         School-Age
4   2015-2016          1227542  ...              Running Late         School-Age

[5 rows x 21 columns]

In [ ]:
# Checking the last 5 rows of data
# ---
#

bb.tail()

School_Year  ...  School_Age_or_PreK
281105   2016-2017  ...               Pre-K
281106   2016-2017  ...               Pre-K
281107   2016-2017  ...          School-Age
281108   2016-2017  ...          School-Age
281109   2016-2017  ...          School-Age

[5 rows x 21 columns]

In [ ]:
# Sample 10 rows of data
# ---
#

bb.sample(10)

School_Year  ...  School_Age_or_PreK
267137   2015-2016  ...          School-Age
132137   2017-2018  ...          School-Age
115855   2018-2019  ...               Pre-K
68915    2016-2017  ...          School-Age
158484   2017-2018  ...          School-Age
206936   2015-2016  ...          School-Age
252275   2017-2018  ...          School-Age
10426    2015-2016  ...          School-Age
231649   2018-2019  ...          School-Age
188344   2017-2018  ...          School-Age

[10 rows x 21 columns]

In [ ]:
# Checking number of rows and columns
# ---
#  
bb.shape

(281110, 21)

In [ ]:
# Checking datatypes
# ---
# 
bb.dtypes

School_Year                        object
Busbreakdown_ID                     int64
Run_Type                           object
Bus_No                             object
Route_Number                       object
Reason                             object
Schools_Serviced                   object
Occurred_On                        object
Created_On                         object
Boro                               object
Bus_Company_Name                   object
How_Long_Delayed                   object
Number_Of_Students_On_The_Bus       int64
Has_Contractor_Notified_Schools    object
Has_Contractor_Notified_Parents    object
Have_You_Alerted_OPT               object
Informed_On                        object
Incident_Number                    object
Last_Updated_On                    object
Breakdown_or_Running_Late          object
School_Age_or_PreK                 object
dtype: object

Record your observations below:

*   The data set contains 281110 rows and 21 columns with most of the values not missing.
*   On a rough glance the values of the Incident_Number column contain, a lot of missing values.
*   Only two columns have integer values, hence arrithemtic operations can be performed on the two only




## 3. External Data Source Validation

You can make sure your data matches other credible information. This ensures that the measurements are roughly in line with what they should be and it serves as a check on what other things might be wrong in your dataset.

External validation can often be as simple as checking your data against some other data i.e. country population your dataset with country population on world bank data.

Replace the text found on this text cell with your external data source validation information.

## 4. Data Preparation

### Performing Data Cleaning

In [ ]:
# Checking datatypes and missing entries of all the variables
# ---
# 

bb.isnull()

School_Year  ...  School_Age_or_PreK
0             False  ...               False
1             False  ...               False
2             False  ...               False
3             False  ...               False
4             False  ...               False
...             ...  ...                 ...
281105        False  ...               False
281106        False  ...               False
281107        False  ...               False
281108        False  ...               False
281109        False  ...               False

[281110 rows x 21 columns]

We observe the following from our dataset:

*   In the above cell, we check if each individual value is check if it is missing.
*   It can be seen that the values of the Incident_Number column contain, a lot of missing values.
*   The method is not ideal, in that it only displays the first and last five records.



In [ ]:
# Standardizing your dataset i.e. variable renaming
# ---
# 
bb.columns = bb.columns.str.strip().str.lower().str.replace(' ', '_')

bb.head()

school_year  busbreakdown_id  ... breakdown_or_running_late school_age_or_prek
0   2015-2016          1227538  ...              Running Late         School-Age
1   2015-2016          1227539  ...              Running Late         School-Age
2   2015-2016          1227540  ...              Running Late              Pre-K
3   2015-2016          1227541  ...              Running Late         School-Age
4   2015-2016          1227542  ...              Running Late         School-Age

[5 rows x 21 columns]

We observe the following from our dataset:

*   From the above cell we can not that, while recording the data, the variable were probaly recorded with no errors to note.
*   In this case we convert the variable names to lower case for easier manipulation of data, while referencing them.


In [ ]:
# Checking how many duplicate rows are there in the data
# ---
# 

sum(bb.duplicated())

#bb_dublicated = bb[bb.duplicated()]

#bb_dublicated.head


0

We observe the following from our dataset:

*   The data set does not contain any duplicate records




In [ ]:
# Checking if any of the columns are all null
# ---
# 

#bb.isnull().sum()---Gives number of null values per column

bb.isnull().any() #This gives a true false result


school_year                        False
busbreakdown_id                    False
run_type                            True
bus_no                              True
route_number                        True
reason                              True
schools_serviced                    True
occurred_on                        False
created_on                         False
boro                                True
bus_company_name                   False
how_long_delayed                    True
number_of_students_on_the_bus      False
has_contractor_notified_schools    False
has_contractor_notified_parents    False
have_you_alerted_opt               False
informed_on                        False
incident_number                     True
last_updated_on                    False
breakdown_or_running_late          False
school_age_or_prek                 False
dtype: bool

We observe the following from our dataset:

*   The data set has 8 Columns with missing values.




In [ ]:
# Checking if any of the rows are all null
# ---
#

# bb.isnull().any(axis = 1) 

bb.isnull().sum(axis = 1)

0         2
1         1
2         1
3         1
4         2
         ..
281105    1
281106    1
281107    0
281108    2
281109    1
Length: 281110, dtype: int64

We observe the following from our dataset:

*   The above shows that most of the records have less than 5 missin values, hence incase of dropping missing values, it is advisable to drop by columns.



In [ ]:
# Checking if the "Yes/No" fields contain only these 2 values
# for have_you_alerted_opt variable
# ---
# Hint: Use unique() function
#que

bb.have_you_alerted_opt.unique()

array(['Yes', 'No'], dtype=object)

We observe the following from our dataset:

*   The column have_you_alerted_opt only contains Yes and No values




In [ ]:
# Checking if the "Yes/No" fields contain only these 2 values
# for has_contractor_notified_parents variable
# ---
# 

bb.has_contractor_notified_parents.unique()

array(['No', 'Yes'], dtype=object)

We observe the following from our dataset:

*   The colomn contain only Yes and No values


In [ ]:
# Checking if the "Yes/No" fields contain only these 2 values
# for has_contractor_notified_schools variable
# ---
#  

bb.has_contractor_notified_schools.unique()


array(['Yes', 'No'], dtype=object)

We observe the following from our dataset:

*   The colomn contain only Yes and No values



In [ ]:
# Checking unique values in breakdown_or_running_late variable to ensure there is no duplication
# ---
#

bb.breakdown_or_running_late.unique()

array(['Running Late', 'Breakdown'], dtype=object)

In [ ]:
# Checking unique values in school_age_or_prek variable
# ---
# 
 
bb.school_age_or_prek.unique()

array(['School-Age', 'Pre-K'], dtype=object)

In [ ]:
# Checking unique values in school_year variable
# ---
#  

bb.school_year.unique()

array(['2015-2016', '2017-2018', '2018-2019', '2016-2017', '2019-2020'],
      dtype=object)

In [ ]:
# Checking unique values in reason variable
# ---
#  

bb.reason.unique()

array(['Heavy Traffic', 'Flat Tire', 'Other', 'Won`t Start',
       'Mechanical Problem', 'Problem Run', 'Accident',
       'Late return from Field Trip', 'Delayed by School',
       'Weather Conditions', nan], dtype=object)

In [ ]:
# Checking unique values in run_type variable
# ---
#  

bb.run_type.unique()

array(['Special Ed AM Run', 'Pre-K/EI', 'General Ed AM Run',
       'General Ed Field Trip', 'Special Ed PM Run', 'General Ed PM Run',
       'Project Read PM Run', 'Special Ed Field Trip',
       'Project Read AM Run', 'Project Read Field Trip', nan],
      dtype=object)

In [ ]:
# Checking unique values in boro variable
# ---
#  

bb.boro.unique()

array(['New Jersey', 'Manhattan', 'Bronx', 'Westchester', 'Brooklyn',
       'Rockland County', 'Nassau County', nan, 'Queens', 'Staten Island',
       'Connecticut', 'All Boroughs'], dtype=object)

### Overall Data Cleaning Observations
**Missing Values**

- There are a large number of missing values in the fields "How_Long_Delayed" which is important to our analysis. 
- There is an extremely large number of missing values in the "Incident_number" field but this is not incidental to our analysis and cannot be filled in without additional information.

**Error in values**

- "How_Long_Delayed" contains string values such as "MINS" or "mins" and a range of values, which needs to be changed to single integer value for our analysis.

**Error in Datatypes**

- "How_Long_Delayed" is a string datatype, should be converted to integer type.

**Error in field names**
- The column name "Boro" should be renamed to "Borough".


### Next Steps: Data Cleaning Steps

**Error in values**

- Extract the first integer value (lowest delay time) in the column "How_Long_Delayed"


**Missing Values**

- Impute the missing values in the field "How_Long_Delayed" with the mean value.


**Error in Datatypes**

- Convert "How_Long_Delayed" to int datatype.



**Error in field names**

- Rename the column "Boro" to "Borough".

In [ ]:
# Lets first start by creating a copy of our dataframe 
# df_clean = df.copy(). We will use this copy as our cleaning copy.
# ---
#
bb_clean = bb.copy()
bb_clean.head()

school_year  busbreakdown_id  ... breakdown_or_running_late school_age_or_prek
0   2015-2016          1227538  ...              Running Late         School-Age
1   2015-2016          1227539  ...              Running Late         School-Age
2   2015-2016          1227540  ...              Running Late              Pre-K
3   2015-2016          1227541  ...              Running Late         School-Age
4   2015-2016          1227542  ...              Running Late         School-Age

[5 rows x 21 columns]

In [ ]:
# Then extracting the lowest delay time in the column how_long_delayed from the string
# 
bb_clean['how_long_delayed'] = bb_clean['how_long_delayed'].str.extract('(\d+)')
bb_clean['how_long_delayed'].head()

0    NaN
1     20
2     15
3     15
4    NaN
Name: how_long_delayed, dtype: object

We impute the null values in 'how_long_delayed' column with mean of the column. This will take a couple of steps...

In [ ]:
# We first convert our how_long_delayed to float type to allow for imputation
# ---
# 

bb_clean["how_long_delayed"] = bb_clean["how_long_delayed"].apply(np.float64)

bb_clean["how_long_delayed"].dtypes

dtype('float64')

In [ ]:
# Then later perform our mean imputation 
# ---
# 

#bb_clean["how_long_delayed"].mean(skipna = True)--- TO find the mean 

# bb_clean["how_long_delayed"] = bb_clean["how_long_delayed"].fillna(np.nan,(bb_clean["how_long_delayed"].mean(skipna = True)))

# .fillna con also be used instead of .replace


bb_clean["how_long_delayed"] = bb_clean["how_long_delayed"].replace(np.nan,(bb_clean["how_long_delayed"].mean(skipna = True)))
bb_clean["how_long_delayed"]

0         28.148313
1         20.000000
2         15.000000
3         15.000000
4         28.148313
            ...    
281105    15.000000
281106    20.000000
281107    20.000000
281108    28.148313
281109    45.000000
Name: how_long_delayed, Length: 281110, dtype: float64

In [ ]:
# Then convert back our how_long_delayed column to integer datatype
# ---
# 

bb_clean["how_long_delayed"] = bb_clean["how_long_delayed"].apply(np.int64)

bb_clean["how_long_delayed"]

#bb_clean["how_long_delayed"].dtypes


0         28
1         20
2         15
3         15
4         28
          ..
281105    15
281106    20
281107    20
281108    28
281109    45
Name: how_long_delayed, Length: 281110, dtype: int64

In [ ]:
# Then check for nulls in the column
# ---
# 
bb_clean["how_long_delayed"].isnull().any()


False

In [ ]:
# Rename Boro column to Borough
# ---
# 

bb_clean.rename(columns={"boro":"borough"}, inplace = True)

bb_clean.head()

school_year  busbreakdown_id  ... breakdown_or_running_late school_age_or_prek
0   2015-2016          1227538  ...              Running Late         School-Age
1   2015-2016          1227539  ...              Running Late         School-Age
2   2015-2016          1227540  ...              Running Late              Pre-K
3   2015-2016          1227541  ...              Running Late         School-Age
4   2015-2016          1227542  ...              Running Late         School-Age

[5 rows x 21 columns]

In [ ]:
# Lastly we convert all values in our colums to lower case
# for ease of reading
# ---
# 
'''
bb_lowercase = bb_clean['school_year', 'busbreakdown_id', 'run_type', 'bus_no', 'route_number',
       'reason', 'schools_serviced', 'occurred_on', 'created_on', 'borough',
       'bus_company_name', 'how_long_delayed', 'number_of_students_on_the_bus',
       'has_contractor_notified_schools', 'has_contractor_notified_parents',
       'have_you_alerted_opt', 'informed_on', 'incident_number',
       'last_updated_on', 'breakdown_or_running_late', 'school_age_or_prek'].str.lower()
'''
#Only values in the borough are converted to lowercase

#bb_clean['borough'] = bb_clean["borough"].str.lower()

bb_lower = bb_clean.applymap(lambda x: x.lower() if type(x)== str else x)
bb_lower.head()


school_year  busbreakdown_id  ... breakdown_or_running_late school_age_or_prek
0   2015-2016          1227538  ...              running late         school-age
1   2015-2016          1227539  ...              running late         school-age
2   2015-2016          1227540  ...              running late              pre-k
3   2015-2016          1227541  ...              running late         school-age
4   2015-2016          1227542  ...              running late         school-age

[5 rows x 21 columns]

In [ ]:
# Check the first 5 record the cleaned dataset
# ---
#
bb_clean = bb_lower
bb_clean.head()

school_year  busbreakdown_id  ... breakdown_or_running_late school_age_or_prek
0   2015-2016          1227538  ...              running late         school-age
1   2015-2016          1227539  ...              running late         school-age
2   2015-2016          1227540  ...              running late              pre-k
3   2015-2016          1227541  ...              running late         school-age
4   2015-2016          1227542  ...              running late         school-age

[5 rows x 21 columns]


## 5. Solution Implementation

Here we investigate the questions that would help craft our recommendations.

### 5.a) Questions

In [ ]:
# 1. Which bus companies that had the highest breakdowns?
# ---
#
breakdowns = bb_clean.groupby(['bus_company_name']).count().reset_index()

# Sort to get bus company with highest breakdowns
# ---
# YOUR CODE GOES BELOW
#
breakdowns.sort_values(by='school_year', ascending=0)


bus_company_name  ...  school_age_or_prek
38                        g.v.c., ltd.  ...               19394
54  leesel transportation corp (b2192)  ...               17200
89           reliant trans, inc. (b232  ...               13741
80           pioneer transportation co  ...               12017
15                  boro transit, inc.  ...               11953
..                                 ...  ...                 ...
51                       l&m bus corp.  ...                   1
36                 fortuna bus company  ...                   1
78                 phillip bus service  ...                   1
87         r & c transit, inc. (b2321)  ...                   1
0                                 1967  ...                   1

[116 rows x 21 columns]

In [ ]:
# 2. What were the top 3 reasons for bus delays?
# ---
#
bus_delay = bb_clean.groupby(['reason']).count()[['how_long_delayed']].reset_index()

# Sort to get the most frequent reason
# ---
# YOUR CODE GOES BELOW
#

bus_delay.sort_values(by ='how_long_delayed', ascending = 0)

#bus_delay

reason  how_long_delayed
3                heavy traffic            173221
6                        other             37579
5           mechanical problem             28162
9                  won`t start             12283
2                    flat tire              8411
8           weather conditions              6932
4  late return from field trip              5704
7                  problem run              4068
0                     accident              2526
1            delayed by school              2222

In [ ]:
# 3. How many students were in the buses when they broke down? 
# ---
#
students = bb_clean.groupby(['number_of_students_on_the_bus']).count()[['busbreakdown_id']].reset_index()

students.sample(20)


number_of_students_on_the_bus  busbreakdown_id
204                           3326                1
53                              53                7
111                            863                1
6                                6             7841
15                              15             1192
213                           5554                1
183                           2557                1
173                           2351                1
217                           5824                1
51                              51                4
8                                8             5402
11                              11             2250
82                             156                1
30                              30              264
24                              24               69
7                                7             6470
160                           2063                1
4                                4            12248
123                           1177                2
112                            871                1

In [ ]:
# 4. Which were most frequent reasons for bus breakdowns?
# ---
#
breakdown_reasons = bb_clean[bb_clean.breakdown_or_running_late == 'breakdown'].groupby(['reason']).count()

# Sort to get most frequent reasons
# ---
# YOUR CODE GOES BELOW
#
breakdown_reasons.sort_values(by="school_year", ascending=0)

school_year  ...  school_age_or_prek
reason                                    ...                    
mechanical problem                 14985  ...               14985
won`t start                         7731  ...                7731
flat tire                           4038  ...                4038
other                               3530  ...                3530
heavy traffic                        419  ...                 419
accident                             202  ...                 202
weather conditions                    82  ...                  82
late return from field trip           35  ...                  35
problem run                           28  ...                  28
delayed by school                      7  ...                   7

[10 rows x 20 columns]

In [ ]:
# 5. What were the most frequent reasons for the bus running late? 
# ---
#

#reasons = bb_clean.groupby(['reason','how_dong_delayed']).count()


# Get the records with running late reasons and sort to get most frequent reasons
# ---
# YOUR CODE GOES BELOW
#

reasons = bb_clean[bb_clean.breakdown_or_running_late == 'running late'].groupby(['reason']).count()[['how_long_delayed']].reset_index()
reasons.sort_values(by="how_long_delayed", ascending=0)

reason  how_long_delayed
3                heavy traffic            172802
6                        other             34049
5           mechanical problem             13177
8           weather conditions              6850
4  late return from field trip              5669
9                  won`t start              4552
2                    flat tire              4373
7                  problem run              4040
0                     accident              2324
1            delayed by school              2215

In [ ]:
# 6. What was the average delay time for each reason type?
# ---
# 

# Get the records with reasons and how long on average a delay took then sort 
# ---
# YOUR CODE GOES BELOW
#
avg_delay = bb_clean.groupby('reason').mean()[['how_long_delayed']].reset_index()

avg_delay.sort_values(by="how_long_delayed", ascending=0)

reason  how_long_delayed
4  late return from field trip         62.235975
0                     accident         34.965162
7                  problem run         31.136676
5           mechanical problem         30.915205
9                  won`t start         29.644061
2                    flat tire         29.392343
8           weather conditions         29.227351
6                        other         28.421006
3                heavy traffic         26.225550
1            delayed by school         18.907741

### 5.b) Recommendations

From the above analysis, below are our recommendations:

1.   
2.   





## 6. Challenging your Solution

During this step, we review our solution and implement  approaches that could potentially provide a better outcome. In our case, we could propose the following question that wasn't answered in our solution because it couldn't have greatly contributed to our recommendation.

In [ ]:
# Which boroughs experienced the most breakdowns?
# ---  
#
breakdowns = bb_clean.groupby(['borough']).count()

# Sort to get bus company with highest breakdowns
# ---
# YOUR CODE GOES BELOW
#

breakdowns.sort_values(by="how_long_delayed", ascending=0)

school_year  ...  school_age_or_prek
borough                       ...                    
bronx                  70952  ...               70952
manhattan              68177  ...               68177
brooklyn               62239  ...               62239
queens                 38060  ...               38060
staten island          13907  ...               13907
westchester             7093  ...                7093
nassau county           4124  ...                4124
new jersey              1583  ...                1583
rockland county          875  ...                 875
all boroughs             429  ...                 429
connecticut              210  ...                 210

[11 rows x 20 columns]

Our observations:

1.   
2.  


How does this observation tie to our solution?



## 7. Follow up questions

During this step, you rethink and propose other ways that you can improve your solution.

### a). Did we have the right data?

### b). Do we need other data to answer our question?

You can look into the questions you brainstormed that you weren't taken into account during analysis due to a lack of data. Were those questions important to have been left out  of your analysis?

### c). Did we have the right question?

Were there any other questions that we needed to have answered?